doc/image-format.md

```diff
@@ -0,0 +1,276 @@
+produces exactly the same output.  However, this document is probably currently
+very incomplete, and maybe even incorrect.  We should strive to cover every
+possible detail here, but it's hard.  Hopefully things will improve with time,
+but until then, you might need to check the implementation.
+
+In cases of ambiguity or incorrectness, issues and patches are extremely
+welcome.
+
+## Overall layout concept
+
+The composefs header and superblock are the only things that need to be at
+fixed offsets.  How do we organize everything else?
+
+Generally speaking, we perform these steps:
+*    collect the filesystem into a flat list of inodes
+*    collect and "share" xattrs, as appropriate
+*    write the composefs header and the superblock
+*    write the inodes directly following the superblock
+*    write the shared xattrs directly following the inodes
+*    then the blocks (only for directories)
+
+## Collecting inodes
+
+We collect the inodes into a flat list according to the following algorithm:
+*   our goal is to visit each inode, collecting it into the inode list as we
+    visit it, in the order that we visited it
+*   start at the root directory
+*   for each directory that we visit:
+    -   the directory is stored first, then the children
+    -   we visit the children in asciibetical order, regardless of file type
+        (ie: we interleave directories and regular files)
+    -   when visiting a child directory, we store all content of the child
+        directory before returning to the parent directory (ie: depth first)
+*   in the case of hardlinks, the inode gets added to the list at the spot that
+    the first link was encountered
+
+Consider a filesystem tree
+
+```
+ /
+   bin/
+     cfsctl
+   usr/
+     lib/
+       libcomposefs.so
+       libglib-2.0.so
+     libexec/
+       cfsctl
+```
+
+where `/bin/cfsctl` and `/usr/libexec/cfsctl` are hardlinks.
+
+In that case, we'd collect the inodes in this order:
+1.  `/`
+1.  `/bin/`
+1.  `/bin/cfsctl` (aka `/usr/libexec/cfsctl`)
+1.  `/usr/`
+1.  `/usr/lib/`
+1.  `/usr/lib/libcomposefs.so`
+1.  `/usr/lib/libglib-2.0.so`
+1.  `/usr/libexec/`
+
+(skipping `/usr/libexec/ctlctl` because we already had it by the time we encountered it).
+
+So that's 8 inodes, in that order.
+
+## Special handling for overlayfs
+
+Ultimately, the erofs image that we produce needs to be used as a layer in an
+overlayfs stack.  There are a lot of cases where the thing that we write out
+only makes sense to overlayfs.  There are other cases where we need to avoiding
+writing out things that overlayfs would treat as "special".
+
+`libcomposefs` writes 256 files named from `00` to `ff` into the root directory
+as character devices with major/minor of (0, 0).  Those are overlayfs whiteouts
+and they are needed for older versions of overlayfs which don't support "data
+only" layers.  We don't target these versions, so *we don't add these files*.
```

Comment:
Here and almost everywhere else...you use "we" to mean this codebase only but I think it'd be more helpful for us to consider "we" to also include the C composefs project.

So we'd rephrase this to something more like "1.1 format doesn't have these files" or so?